rosetta-core/src/node_uri.rs

```diff
@@ -0,0 +1,176 @@
+use fluent_uri::Uri;
+use std::fmt::{Display, Formatter};
+use thiserror::Error;
+
+#[derive(Debug, Error, PartialEq)]
+pub enum NodeUriError {
+    #[error("Invalid uri")]
+    InvalidUri,
+    #[error("Invalid uri scheme")]
+    InvalidScheme,
+    #[error("Invalid uri port")]
+    InvalidPort,
+    #[error("Invalid uri host")]
+    InvalidHost,
+}
+
+/// The URI is structured as follows:
+///
+/// ```notrust
+/// abc://username:password@example.com:1234/path/data?key=value&key2=value2#fragid1
+/// |-|  |----------------| |---------| |--||--------| |-------------------| |-----|
+///  |            |              |       |      |               |              |
+/// scheme    userinfo         host    port   path            query         fragment
+/// ```
+#[derive(Debug, Clone, PartialEq)]
+pub struct NodeUri<'a> {
+    pub scheme: &'a str,
+    pub userinfo: Option<&'a str>,
+    pub host: &'a str,
+    pub port: u16,
+    pub path: &'a str,
+    pub query: Option<&'a str>,
+    pub fragment: Option<&'a str>,
+}
```

Comment:
I've created our own Uri wrapper, I could use one some one existing libraries like [http::uri](https://docs.rs/http/latest/http/uri/index.html) or [uriparse-rs](https://github.com/sgodwincs/uriparse-rs). But we have some extra requirements such as the `scheme` and `port` are mandatory, required for rosetta-docker. Also once this is a core struct, I think is better use our own struct, so we have more control in case of breaking changes.

I also decided to use [fluent-uri](https://github.com/yescallop/fluent-uri-rs) instead [uriparse-rs](https://github.com/sgodwincs/uriparse-rs) because fluent-uri have a better performance, is being actively maintained and support `no_std` environments.